bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ProjectEncodingPreferenceChangeListener.java

```diff
@@ -0,0 +1,198 @@
+/*******************************************************************************
+ * Copyright (c) 2022 IBM Corporation and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Ingo Mohr - Issue #166 - Add Preference to Turn Off Warning-Check for Project Specific Encoding
+ *******************************************************************************/
+package org.eclipse.core.internal.resources;
+
+import java.util.*;
+import org.eclipse.core.internal.utils.Messages;
+import org.eclipse.core.resources.*;
+import org.eclipse.core.runtime.*;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences.IPreferenceChangeListener;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences.PreferenceChangeEvent;
+import org.eclipse.core.runtime.preferences.IScopeContext;
+import org.osgi.service.component.annotations.*;
+
+@Component(service = {})
+public class ProjectEncodingPreferenceChangeListener implements IPreferenceChangeListener {
+
+	private static final String MARKER_TYPE = ValidateProjectEncoding.MARKER_TYPE;
+
+	@Reference
+	private IWorkspace workspace;
+
+	@Reference
+	private ILog log;
+
+	@Reference(target = IScopeContext.BUNDLE_SCOPE_FILTER)
+	IScopeContext bundleScope;
+
+	private IEclipsePreferences eclipsePreferences;
+
+	@Activate
+	public void register() {
+		eclipsePreferences = bundleScope.getNode(""); //$NON-NLS-1$
```

Comment:
Is this the same node as this later?

```
IEclipsePreferences node = InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES);
```

Reply:
The preferences are already available as `private IEclipsePreferences eclipsePreferences;` so no need to do this static stuff here.

Reply:
> The preferences are already available as `private IEclipsePreferences eclipsePreferences;` so no need to do this static stuff here.

I don't understand this comment.  Where/how is the value assigned?  How is equivalent to the node `InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES)`?

Reply:
I figured they must be the same node, otherwise the code wouldn't work, but it's certainly not obviously the case.

You mention there is  "no need to call later in static parts" and again, forgive me for being dense, but where exactly do you mean is that other place?  Note that the call to `InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES)` in ValidateProjectEncoding.java which is a different file that does not have access to the field  eclipsePreferences in this preference listener class.   So in which class are you suggesting a change?  Or are you not suggesting a change but only answering the "are the nodes equal" question?

Reply:
> ValidateProjectEncoding.java which is a different file that does not have access to the field eclipsePreferences in this preference listener class.

Okay I missed this then forget about that, still they are the same :-)

> Or are you not suggesting a change but only answering the "are the nodes equal" question?

Yes, even though `ValidateProjectEncoding` might also be converted to a component we should not mix that here then.

> but it's certainly not obviously the case.

This is a quite new feature, the `org.eclipse.core.runtime.preferences.IScopeContext.TYPE_BUNDLE` explains this and if you take a look at `org.eclipse.core.internal.preferences.BundleStateScope` you will see that under the hood it calls `InstanceScope.INSTANCE.getNode(<id of bundle>)` and `ResourcesPlugin.PI_RESOURCES` is a constant to the resource bundle id.

Reply:
Thanks for the feedback. I'm also for not mixing things up for ``ValidateProjectEncoding`` (at least not for this PR)